pip/src/noisy_simulator.rs

```diff
@@ -0,0 +1,367 @@
+    }
+
+    pub fn get_operation_matrix(&self) -> Vec<Complex<f64>> {
+        nalgebra_matrix_to_python_list(self.0.matrix())
+    }
+
+    pub fn get_kraus_operators(&self) -> Vec<Vec<Complex<f64>>> {
+        let mut kraus_operators = Vec::new();
+        for kraus_operator in self.0.kraus_operators() {
+            kraus_operators.push(nalgebra_matrix_to_python_list(kraus_operator));
+        }
+        kraus_operators
+    }
+
+    pub fn get_number_of_qubits(&self) -> usize {
+        self.0.number_of_qubits()
+    }
+}
+
+#[pyclass]
+pub(crate) struct Instrument(noisy_simulator::Instrument);
+
+#[pymethods]
+impl Instrument {
+    #[new]
+    pub fn new(operations: Vec<Operation>) -> Self {
+        let operations = operations.into_iter().map(|op| op.0).collect();
+        Self(noisy_simulator::Instrument::new(operations))
+    }
+}
+
+#[pyclass]
+#[derive(Clone)]
+pub(crate) struct DensityMatrix {
+    /// Dimension of the matrix. E.g.: If the matrix is 5 x 5, then dimension is 5.
+    dimension: usize,
+    /// Number of qubits in the system.
+    number_of_qubits: usize,
+    /// Theoretical change in trace due to operations that have been applied so far.
+    trace_change: f64,
+    /// Vector storing the entries of the density matrix.
+    data: Vec<Complex<f64>>,
+}
+
+#[pymethods]
+impl DensityMatrix {
+    /// Returns a copy of the matrix data.
+    fn data(&self) -> Vec<Complex<f64>> {
+        self.data.clone()
+    }
+
+    /// Returns the dimension of the matrix.
+    fn dimension(&self) -> usize {
+        self.dimension
+    }
+
+    /// Returns the number of qubits in the system.
+    fn number_of_qubits(&self) -> usize {
+        self.number_of_qubits
+    }
+}
+
+impl From<&noisy_simulator::DensityMatrix> for DensityMatrix {
+    fn from(dm: &noisy_simulator::DensityMatrix) -> Self {
+        Self {
+            dimension: dm.dimension(),
+            number_of_qubits: dm.number_of_qubits(),
+            trace_change: dm.trace_change(),
+            data: dm.data().iter().copied().collect::<Vec<_>>(),
+        }
+    }
+}
+
+impl TryInto<noisy_simulator::DensityMatrix> for DensityMatrix {
+    type Error = PyErr;
+
+    fn try_into(self) -> PyResult<noisy_simulator::DensityMatrix> {
+        noisy_simulator::DensityMatrix::try_from(
+            self.dimension,
+            self.number_of_qubits,
+            self.trace_change,
+            ComplexVector::from_vec(self.data),
+        )
+        .ok_or(NoisySimulatorError::new_err("invalid density matrix"))
+    }
+}
+
+#[pyclass]
+pub(crate) struct DensityMatrixSimulator(noisy_simulator::DensityMatrixSimulator);
+
+#[pymethods]
+impl DensityMatrixSimulator {
+    #[new]
+    pub fn new(number_of_qubits: usize, seed: Option<u64>) -> Self {
+        let mut simulator = noisy_simulator::DensityMatrixSimulator::new(number_of_qubits);
+
+        if let Some(seed) = seed {
+            simulator.set_rng_seed(seed);
+        }
```

Comment:
Should we pass this seed into the new fn instead of calling set_rng_seed manually?

Reply:
I am not sure. Do you mean passing an `Option<u64>` to `new`? Rather than having an argument that is `None` some of the times, I would preffer having `fn with_rng_seed(self, seed: u64) -> Self` such that we can do:

```rust
let sim = DensityMatrixSimulator::new(number_of_qubits)
    .with_rng_seed(seed);
```

Reply:
an alternative option could be to have `new` stay as is, taking only number of qubits, and a second constructor `new_with_seed` that takes an optional seed. Then the first `new` is just a thin wrapper around `new_with_seed` and majority of callers can avoid having to pass explicit `None`.

Reply:
Yes, I like that alternative, because what I suggest is usually used when you have many optional arguments.